### PR TITLE
[10.0][FIX] stock_valued_picking_report: Fix report

### PR DESCRIPTION
**Pull request by luismontalba**
This PR fix template and refactor code to calculate picking value.

cc @Tecnativa

### CODE REVIEW

**Comment by luismontalba** [comment]
Last commit fix calculation of valued picking when there are several sale lines with same product.

File: stock_valued_picking_report/models/stock_pack_operation.py [hunk 1]
```diff
@@ -53,42 +54,41 @@ def _compute_sale_order_line_fields(self):
...
+        for sale_line in sale_lines:
+            sum_qty += sale_line.product_uom_qty
+            sum_price += sale_line.price_unit * sale_line.product_uom_qty
+            sum_discount += sale_line.discount * sale_line.product_uom_qty
+            amount_untaxed = sale_line.price_subtotal
+            sum_amount_untaxed += amount_untaxed
+            rm = sale_line.order_id.company_id.tax_calculation_rounding_method
+            if rm == 'round_globally':
+                taxes = sale_line.tax_id.compute_all(
+                    sale_line.price_reduce,
```
Comment (pedrobaeza) [inline stock_valued_picking_report/models/stock_pack_operation.py]: This field doesn't exist nor it's computed before.

**Review by pedrobaeza — CHANGES_REQUESTED** [review]

**Comment by luismontalba** [comment]
Changes done and test added.

**Review by pedrobaeza — APPROVED** [review]
@carlosdauden please review

File: stock_valued_picking_report/models/stock_pack_operation.py [hunk 2]
```diff
@@ -53,42 +54,43 @@ def _compute_sale_order_line_fields(self):
 
     @api.multi
     def sale_lines_values(self, sale_lines):
-        if len(sale_lines) <= 1:
-            price_unit = sale_lines.price_unit
```
Comment (carlosdauden) [inline stock_valued_picking_report/models/stock_pack_operation.py]: ¿Do you need reprocess all amounts with one line?

File: stock_valued_picking_report/models/stock_pack_operation.py [hunk 3]
```diff
@@ -53,42 +54,43 @@ def _compute_sale_order_line_fields(self):
...
+            rm = sale_line.order_id.company_id.tax_calculation_rounding_method
+            if rm == 'round_globally':
+                price = sale_line.price_unit * (1 - (sale_line.discount or
+                                                     0.0) / 100)
+                taxes = sale_line.tax_id.compute_all(
+                    price,
+                    sale_line.order_id.currency_id,
+                    sale_line.product_uom_qty,
+                    product=sale_line.product_id,
+                    partner=sale_line.order_id.partner_shipping_id
```
Comment (carlosdauden) [inline stock_valued_picking_report/models/stock_pack_operation.py]: ¿Why partner_shipping_id?

File: stock_valued_picking_report/models/stock_pack_operation.py [hunk 4]
```diff
@@ -53,42 +54,43 @@ def _compute_sale_order_line_fields(self):
...
+            sum_discount += sale_line.discount * sale_line.product_uom_qty
+            amount_untaxed = sale_line.price_subtotal
+            sum_amount_untaxed += amount_untaxed
+            rm = sale_line.order_id.company_id.tax_calculation_rounding_method
+            if rm == 'round_globally':
+                price = sale_line.price_unit * (1 - (sale_line.discount or
+                                                     0.0) / 100)
+                taxes = sale_line.tax_id.compute_all(
+                    price,
+                    sale_line.order_id.currency_id,
```
Comment (carlosdauden) [inline stock_valued_picking_report/models/stock_pack_operation.py]: https://github.com/OCA/OCB/blob/179de9b55f4d3e6006cdbfaa77f67ef4956d0628/addons/sale/models/sale.py#L775

File: stock_valued_picking_report/models/stock_pack_operation.py [hunk 5]
```diff
@@ -53,42 +54,43 @@ def _compute_sale_order_line_fields(self):
...
+                    sale_line.product_uom_qty,
```
Comment (carlosdauden) [inline stock_valued_picking_report/models/stock_pack_operation.py]: What happens if the order is not served completely?

File: stock_valued_picking_report/models/stock_pack_operation.py [hunk 6]
```diff
@@ -53,42 +54,43 @@ def _compute_sale_order_line_fields(self):
...
+        sum_amount_untaxed = 0.0
+        sum_amount_tax = 0.0
+        sum_amount_total = 0.0
+        for sale_line in sale_lines:
+            sum_qty += sale_line.product_uom_qty
+            sum_price += sale_line.price_unit * sale_line.product_uom_qty
+            sum_discount += sale_line.discount * sale_line.product_uom_qty
+            amount_untaxed = sale_line.price_subtotal
+            sum_amount_untaxed += amount_untaxed
+            rm = sale_line.order_id.company_id.tax_calculation_rounding_method
```
Comment (carlosdauden) [inline stock_valued_picking_report/models/stock_pack_operation.py]: Do you need set this inside of the loop?

**Review by carlosdauden — CHANGES_REQUESTED** [review]

File: stock_valued_picking_report/models/stock_pack_operation.py (hunk 6 above)
Comment (luismontalba) [reply stock_valued_picking_report/models/stock_pack_operation.py]: The function take a set of sale order lines as an argument, I think this way let the function deal with sale order lines from different orders with different companies, but any suggestion is welcome :smile:

File: stock_valued_picking_report/models/stock_pack_operation.py (hunk 5 above)
Comment (luismontalba) [reply stock_valued_picking_report/models/stock_pack_operation.py]: In that case the value is not computed partially, I can try to use qty_delivered instead.

File: stock_valued_picking_report/models/stock_pack_operation.py (hunk 6 above)
Comment (luismontalba) [reply stock_valued_picking_report/models/stock_pack_operation.py]: Certainly it's not necessary so I put it outside to improve performance.

**Comment by luismontalba** [comment]
@carlosdauden , changes done

File: stock_valued_picking_report/report/stock_picking_valued_report.xml [hunk 7]
```diff
@@ -13,7 +13,8 @@
                 <th class="text-right"><strong>Taxes</strong></th>
             </t>
         </xpath>
-        <xpath expr="//span[@t-field='pack_operation.product_uom_id']/.." position="after">
+        <xpath expr="//span[@t-field='pack_operation.qty_done_uom_ordered']/."
```
Comment (carlosdauden) [inline stock_valued_picking_report/report/stock_picking_valued_report.xml]: Don't work
captura de pantalla_2017-10-20_09-45-32
captura de pantalla_2017-10-20_09-44-06

**Review by carlosdauden — CHANGES_REQUESTED** [review]

**Comment by luismontalba** [comment]
It works when I try in runbot, even in a new private window in my browser:

imagen

File: stock_valued_picking_report/report/stock_picking_valued_report.xml (hunk 7 above)
Comment (carlosdauden) [reply stock_valued_picking_report/report/stock_picking_valued_report.xml]: As we said yesterday, the delivery note should be able to print before validating.
Why do you remove one dot?

**Comment by luismontalba** [comment]
@carlosdauden , please try now

File: stock_valued_picking_report/report/stock_picking_valued_report.xml [hunk 8]
```diff
@@ -13,7 +13,8 @@
                 <th class="text-right"><strong>Taxes</strong></th>
             </t>
         </xpath>
-        <xpath expr="//span[@t-field='pack_operation.product_uom_id']/.." position="after">
+        <xpath expr="//tr[@t-foreach='o.pack_operation_ids']"
```
Comment (carlosdauden) [inline stock_valued_picking_report/report/stock_picking_valued_report.xml]: Ok, although the previous alternatives access a more specific position:

` <xpath expr="//span[@t-field='pack_operation.qty_done_uom_ordered']/.."`
or 
`<xpath expr="//span[@t-field='pack_operation.product_uom_id'][last()]/.."`

**Review by carlosdauden — APPROVED** [review]

File: stock_valued_picking_report/report/stock_picking_valued_report.xml (hunk 8 above)
Comment (luismontalba) [reply stock_valued_picking_report/report/stock_picking_valued_report.xml]: Very interesting ...thanks!

Comment (pedrobaeza) [reply stock_valued_picking_report/report/stock_picking_valued_report.xml]: Please change it to a more specific one as indicated, or the order can change.

Comment (luismontalba) [reply stock_valued_picking_report/report/stock_picking_valued_report.xml]: Changes done

Comment (pedrobaeza) [reply stock_valued_picking_report/report/stock_picking_valued_report.xml]: But the position is still "inside". Can this be after or before an element for not depending on the inheritance for the order?

Comment (luismontalba) [reply stock_valued_picking_report/report/stock_picking_valued_report.xml]: Yes, last commit better referenced with an "after"

**Comment by pedrobaeza** [comment]
@carlosdauden can you check this last change?

File: stock_valued_picking_report/report/stock_picking_valued_report.xml [hunk 9]
```diff
@@ -13,7 +13,9 @@
                 <th class="text-right"><strong>Taxes</strong></th>
             </t>
         </xpath>
-        <xpath expr="//span[@t-field='pack_operation.product_uom_id']/.." position="after">
+        <xpath
+                expr="//span[@t-field='pack_operation.qty_done_uom_ordered']/."
```
Comment (carlosdauden) [inline stock_valued_picking_report/report/stock_picking_valued_report.xml]: Dot missed

**Review by carlosdauden — CHANGES_REQUESTED** [review]

File: stock_valued_picking_report/report/stock_picking_valued_report.xml (hunk 9 above)
Comment (carlosdauden) [reply stock_valued_picking_report/report/stock_picking_valued_report.xml]: We are returned to first commit 45 day ago:
https://github.com/OCA/stock-logistics-reporting/pull/38/commits/41a15ebc2d3c101b2537c5afacc1d0f9cca9d825

Read my last month comment https://github.com/OCA/stock-logistics-reporting/pull/38/commits/41a15ebc2d3c101b2537c5afacc1d0f9cca9d825#r145897719

**Comment by carlosdauden** [comment]
ping @rafaelbn @pedrobaeza

**Comment by rafaelbn** [comment]
Tested in runbot 👍

**Review by rafaelbn — APPROVED** [review]
👍 
Still missing your approval @carlosdauden

**Comment by carlosdauden** [comment]
I already approved:
https://github.com/OCA/stock-logistics-reporting/pull/38#pullrequestreview-70817228

It would also be ugly for me to approve my changes:
https://github.com/OCA/stock-logistics-reporting/pull/38/commits/2c61611c90f867cd96c64dc43949d7912f847d90
:wink:

**Review by pedrobaeza — APPROVED** [review]

**Pull request merged**